src/middlewared/middlewared/plugins/kmip/sed_keys.py

```diff
@@ -0,0 +1,239 @@
+            try:
+                if disk['passwd']:
+                    key = self.middleware.call_sync('pwenc.decrypt', disk['passwd'])
+                elif self.disks_keys.get(disk['identifier']):
+                    key = self.disks_keys[disk['identifier']]
+                elif connection_successful:
+                    with self._connection(self.middleware.call_sync('kmip.connection_config')) as conn:
+                        key = self._retrieve_secret_data(disk['kmip_uid'], conn)
+                else:
+                    raise Exception('Failed to sync disk')
+            except Exception:
+                failed.append(disk['identifier'])
+            else:
+                update_data = {'passwd': self.middleware.call_sync('pwenc.encrypt', key), 'kmip_uid': None}
+                self.middleware.call_sync(
+                    'datastore.update', self.disks_datastore, disk['identifier'], update_data, {'prefix': 'disk_'}
+                )
+                self.disks_keys.pop(disk['identifier'], None)
+                if connection_successful:
+                    self.middleware.call_sync('kmip.delete_kmip_secret_data', disk['kmip_uid'])
+        adv_config = self.middleware.call_sync('kmip.system_advanced_config')
+        if adv_config['kmip_uid']:
+            key = None
+            if adv_config['sed_passwd']:
+                key = adv_config['sed_passwd']
+            elif self.global_sed_key:
+                key = self.global_sed_key
+            elif connection_successful:
+                try:
+                    with self._connection(self.middleware.call_sync('kmip.connection_config')) as conn:
+                        key = self._retrieve_secret_data(adv_config['kmip_uid'], conn)
+                except Exception:
+                    failed.append('Global SED Key')
+            if key:
+                self.middleware.call_sync(
+                    'datastore.update', self.sys_adv_datastore,
+                    adv_config['id'], {
+                        'adv_sed_passwd': self.middleware.call_sync('pwenc.encrypt', key), 'adv_kmip_uid': None
+                    }
+                )
+                self.global_sed_key = ''
+                if connection_successful:
+                    self.middleware.call_sync('kmip.delete_kmip_secret_data', adv_config['kmip_uid'])
+        return failed
+
+    @job(lock=lambda args: f'kmip_sync_sed_keys_{args}')
+    @private
+    def sync_sed_keys(self, job, ids=None, force=False):
+        if not force and not self.middleware.call_sync('kmip.sed_keys_pending_sync'):
+            return
+        config = self.middleware.call_sync('kmip.config')
+        conn_successful = self.middleware.call_sync('kmip.test_connection', None, True)
+        if config['enabled'] and config['manage_sed_disks']:
+            if conn_successful:
+                failed = self.push_sed_keys(ids)
+            else:
```

Comment:
Should we issue an alert if connection failed?

Reply:
We do raise it above `kmip. test_connection`, `True` there specifies that an alert should be raised